### PR TITLE
[JUJU-189] Raft lease extension timeout

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following changes allow extensions on leases to have a larger
timeout when attempting to enqueue on to the raft lease queue. The aim
behind this change is to allow less churn for a given application lease.

During load testing, it is seen that when claims block the queue all
subsequent extensions are also thrown away. This is highly problematic
as it causes application leader changes, which also cause more churn in
the database. The change here is to be more forgiving for extensions and
hope after a lost claim, we get to push some extensions through so we
can handle more throughput.

## QA steps

See for Q&A steps for a larger in-depth test https://github.com/juju/juju/pull/13453

```
$ juju bootstrap lxd test1 --build-agent --config features="[raft-api-leases,raft-batch-fsm]"
$ juju model-config -m controller logging-config="<root>=INFO;juju.worker.lease.raft=TRACE;juju.core.raftlease=TRACE;juju.worker.globalclockupdater.raft=TRACE;juju.worker.raft=TRACE"
$ juju debug-log -m controller
```

Ensure there are no errors.

### CODE REVIEW

File: api/raftlease/raftlease.go [hunk 1]
```diff
@@ -32,10 +32,11 @@ func NewAPI(caller base.APICaller) *API {
 // ApplyLease attempts to apply a lease against the given controller. If the
 // controller is not the leader, then an error to redirect to a new leader will
 // be given.
-func (api *API) ApplyLease(command string) error {
+func (api *API) ApplyLease(leaseType, command string) error {
```
Comment (manadart) [inline api/raftlease/raftlease.go:35]: Are the comments above in this file regarding the pubsub API correct?

**Comment by manadart** [comment]
I think rather than change the facade to include the type of operation, we should just be passing a serialised `raftlease.Command` in `params` form.

At the moment we do: command -> string -> params -> wire -> params.
We're not introducing a whole lot of overhead to do: command -> params -> wire -> params -> string.

Then we can just interrogate the type member instead of passing it separately to the stringified command.

It is then a short step to prioritise lease name-spaces as we have discussed (if desired).

File: apiserver/params/params.go [hunk 2]
```diff
@@ -1080,6 +1080,49 @@ type LeaseOperation struct {
 	Command string `json:"command"`
 }
 
+// LeaseOperationsV2 is used to send raft operational messages between
+// controllers.
+type LeaseOperationsV2 struct {
+	Operations []LeaseOperationCommand `json:"commands"`
+}
+
+// LeaseOperation is used to send raft operational messages between controllers.
```
Comment (manadart) [inline apiserver/params/params.go:1089]: "LeaseOperationCommand"

File: core/raft/queue/opqueue.go [hunk 3]
```diff
@@ -17,6 +18,12 @@ const (
 	// returned.
 	EnqueueTimeout time.Duration = time.Second * 3
```
Comment (manadart) [inline core/raft/queue/opqueue.go:19]: I'd now rename this to `EnqueueDefaultTimeout`.

**Comment by manadart** [comment]
I really like this. Much better.

**Review by manadart — APPROVED** [review]
I really like this. Much better.

**Comment by SimonRichardson** [comment]
$$merge$$

I'll do a follow up PR this week to see if we're correct with the timeout durations.

**Comment by SimonRichardson** [comment]
!!build!!

**Comment by hpidcock** [comment]
!!build!!

**Comment by SimonRichardson** [comment]
$$merge$$

**Comment by SimonRichardson** [comment]
$$merge$$

**Pull request merged**